### PR TITLE
5 implement mmd

### LINKED ISSUE #5: Implement Maximum Mean Discrepancy
Implement MMD between target-surrogate pairs.
- Some intuition: https://jejjohnson.github.io/research_journal/appendix/similarity/mmd/
- An implementation with a linear kernel: https://github.com/jindongwang/transferlearning/blob/master/code/distance/mmd_numpy_sklearn.py

### PR DESCRIPTION
**Pull request by philswatton**
This PR implements the following:

- `mmd` function which takes two NP arrays, an embedding name, and a kernel name as input
  - internal dicts are used to transform strings to embedding functions and kernel functions
  - this is for compatability with using configs
- `load_config` to handle loading configs given a file path
  - atm it only loads a config for similarity metrics, but is set up for easy extension to e.g. dataset and experiment configs in the future
- tests for both of these, which use their own config file

This resolves #5 and contributes to #19.

I've also opened two new issues to further extend this initial implementation:
- #20 suggests looking at implementing MMD by class (rather than MMD over the whole dataset) in the future, with arguments to handle this to be added to both config and `mmd`
- #21 suggests looking at using feature embeddings as an additional hyperparameter to the MDD

EDIT: I also added some CIFAR bits to the .gitignore as suggested by @jack89roberts at some point somewhere

### CODE REVIEW

File: src/modsim2/similarity/mmd.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        kernel_name: An appropriate kernel embedding. See kernel_dict
+                     for choices
+
+    Returns:
+        float: The maximum mean discrepancy between A and B
+    """
+    # Extract embedding callable
+    embedding = embedding_dict[embedding_name]
+
+    # In the future, consider optionally using feature embeddings here
```
Comment (jack89roberts) [inline src/modsim2/similarity/mmd.py:61]: I think best practice is generally not to have comments like this in the code (or at least to cite an issue number, if keeping it)

File: tests/test_similarity.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir))
+
+# Get metric config
+METRICS_CONFIG_PATH = os.path.join(PROJECT_ROOT, "tests", "testconfig", "metrics.yaml")
+CONFIGS = load_configs(METRICS_CONFIG_PATH)
+METRIC_CONFIG = CONFIGS["metric_config"]
+
+
+# same structure as CIFAR10 but doesn't require a download
+class DummyCIFAR10(CIFAR10):
```
Comment (jack89roberts) [inline tests/test_similarity.py:29]: If these dummy/mock classes are the same as what's used in `test_data.py` they could be pulled out and stored in a common file (`pytest` has a place you're meant to put stuff like that, I forget what it is, maybe `conftest.py`)

**Review by jack89roberts — COMMENTED** [review]
Few initial comments whilst I wait for some things to run 🙂

File: src/modsim2/similarity/mmd.py (hunk 1 above)
Comment (lannelin) [reply src/modsim2/similarity/mmd.py:61]: agreed, tie to an issue or drop

File: tests/test_similarity.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    dataset_cls = DummyCIFAR10
+
+
+@pytest.fixture(scope="module", autouse=True)
+def patch_datamodule():
+    with patch("modsim2.data.loader.CIFAR10DataModule", new=MockCIFAR10DataModule):
+        yield None
+
+
+def test_cifar_mmd_same():
```
Comment (lannelin) [inline tests/test_similarity.py:52]: this module should probably be `test_mmd` and only test mmd related stuff like this method

File: src/modsim2/similarity/mmd.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        np.ndarray: An np.ndarray with the same number of observations
+                    reshaped to a matrix
+    """
+    shape = array.shape
+    matrix = array.reshape(shape[0], np.prod(shape[1:]))
+    return matrix
+
+
+# Dictionary mapping from embedding names to Callables
+embedding_dict = {"matrix": array_to_matrix}
```
[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/similarity/mmd.py:24]: once extracted, with `array_to_matrix`, would also be good to have a separate test module for mmd related stuff and then embedding fns. So we'd have  `similarity/mmd.py` and `similarity/embeddings.py` and then `tests/similarity/test_mmd.py` and `tests/similarity/test_embeddings.py`. Not fussy about exact naming or structure.

**Review by lannelin — CHANGES_REQUESTED** [review]
Looking good so far! Some fairly minor changes requested in comments.

File: src/modsim2/data/loader.py [hunk 5]
```diff
@@ -347,3 +351,26 @@ def __init__(
...
+    def compute_similarity(self):
+        data_A = self.A.dataset_train.dataset.data[
+            self.indices_A,
+        ]
+        data_B = self.B.dataset_train.dataset.data[
+            self.indices_B,
+        ]
+
+        # Manages calling of functions from config string input
+        function_dict = {"mmd": mmd}
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/data/loader.py:365]: I've done this

[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/data/loader.py:365]: Done

File: src/modsim2/data/loader.py [hunk 6]
```diff
@@ -347,3 +351,26 @@ def __init__(
...
+
+        # Loop over dict, compute metrics
+        similarity_dict = {}
+        config_keys = self.metric_config.keys()
+        for key in config_keys:
+            metric = self.metric_config[key]
+            similarity_dict[key] = function_dict[metric["function"]](
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/data/loader.py:370]: I'm unsure atm - I'm currently working on the assumption that when we calculate the metrics we'll want to immediately write them to an output, then merge that output later on with an output in terms of attack transfer success. Once we've computed them though I don't know of any reason we'd use them beyond this purpose

File: src/modsim2/similarity/mmd.py (hunk 1 above)
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:61]: I've moved the embeddings stuff to its own file w/out a comment re this

File: src/modsim2/similarity/mmd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:24]: these are now in their own similarity.embeddings module. Still need to add the embeddings-specific tests

File: src/modsim2/similarity/mmd.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        array_A: The first image dataset
+        array_B: The second image dataset
+        kernel_name: An appropriate kernel embedding. See kernel_dict
+                     for choices
+
+    Returns:
+        float: The maximum mean discrepancy between A and B
+    """
+    # Extract embedding callable
+    embedding = embedding_dict[embedding_name]
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:59]: done

[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:59]: with kernel also becoming kernel_fn

File: tests/test_similarity.py (hunk 3 above)
Comment (philswatton) [reply tests/test_similarity.py:52]: done

File: src/modsim2/data/loader.py [hunk 8]
```diff
@@ -347,3 +351,26 @@ def __init__(
...
+        ]
+
+        # Manages calling of functions from config string input
+        function_dict = {"mmd": mmd}
+
+        # Loop over dict, compute metrics
+        similarity_dict = {}
+        config_keys = self.metric_config.keys()
+        for key in config_keys:
+            metric = self.metric_config[key]
```
[the remark this answers is not in the document]
Comment (jack89roberts) [reply src/modsim2/data/loader.py]: oops looks like I wrote the wrong variable name here @philswatton -  should have been `for key, metric in self.metric_config.items()` I think - sorry!

[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/data/loader.py]: yep - I caught it though, dw!

File: tests/test_similarity.py (hunk 2 above)
Comment (philswatton) [reply tests/test_similarity.py:29]: This is done!

File: src/modsim2/similarity/mmd.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import numpy as np
+from sklearn import metrics
+
+
+def array_to_matrix(array: np.ndarray) -> np.ndarray:
+    """
+    Reshapes arbitrarily sized np.ndarrays to a matrix with the
```
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py]: Done

[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py]: And yes, I think so - even if the size of the second dim is only 1

File: src/modsim2/similarity/mmd.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (philswatton) [reply src/modsim2/similarity/mmd.py]: I've now added a test for the embeddings

File: src/modsim2/similarity/mmd.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (lannelin) [reply src/modsim2/similarity/mmd.py]: I think the -1 actually solves this for you and you *can* pass a 1d array in. Is this expected? if not add a guard that prevents it like:
```
if array.ndim < 2:
    raise Exception("blah blah")
```

File: src/modsim2/similarity/mmd.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+
+    Note that A and B must be matrices for the MMD. Currently mmd()
+    will reshape arbitarily size np.ndarrays to matrices. However,
+    other options include feature embeddings which will be explored in
+    the future.
+
+    Args:
+        array_A: The first image dataset
+        array_B: The second image dataset
+        kernel_name: An appropriate kernel embedding. See kernel_dict
```
Comment (lannelin) [inline src/modsim2/similarity/mmd.py:30]: missing embedding_name here

File: src/modsim2/similarity/mmd.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+    matrix_A = embedding_fn(array_A)
+    matrix_B = embedding_fn(array_B)
+    N_A = matrix_A.shape[0]
+    N_B = matrix_B.shape[0]
+
+    # Extract kernel callable
+    kernel_fn = MMD_KERNEL_DICT[kernel_name]
+
+    # Compute MMD components
+    kernel_AA = kernel_fn(X=matrix_A)
```
Comment (lannelin) [inline src/modsim2/similarity/mmd.py:49]: these ops tend to be expensive and looks like the sklearn ones don't provide their own logging. I think it'd be helpful to add log statements here (probably at INFO) to state we're about to compute kernel_fn [name] over N pairs of rows and that this may be a bit slow

**Review by lannelin — COMMENTED** [review]
looks great, a couple of very minor comments then happy to merge without re-review.

Have run the following on my machine and returns a very small value for mmd (`1.5625e-05`) with rbf kernel. I imagine this is somewhat expected even with fairly large drop sizes?
```
from modsim2.data.loader import DMPair

metric_config = {'mmd_rbf': {'function': 'mmd', 'arguments': {'embedding_name': 'matrix', 'kernel_name': 'rbf'}}}

dmpair = DMPair(metric_config=metric_config, drop_percent_A=0.2, drop_percent_B=0.2)

dmpair.compute_similarity()
```

File: src/modsim2/similarity/mmd.py (hunk 10 above)
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:38]: I've added this

And yes - given the way MMD works and the fact that we stratify our drops by class, I don't think dropping should change this (or probably a lot of the other metrics too - though PAD might be vulnerable to it)

File: src/modsim2/similarity/mmd.py (hunk 11 above)
Comment (philswatton) [reply src/modsim2/similarity/mmd.py:72]: I've added this

File: src/modsim2/similarity/mmd.py [hunk 12]
```diff
@@ -46,6 +54,21 @@ def mmd(
...
+        + str(N_A)
+        + ", "
+        + str(N_B)
+        + " by "
+        + str(N_B)
+        + ", and "
+        + str(N_A)
+        + " by "
+        + str(N_B)
+        + " kernels. This may take some time!"
```
Comment (jack89roberts) [inline src/modsim2/similarity/mmd.py:70]: Using f-strings would be more canonical, e.g.:

```suggestion
        f"Computing {N_A} by {N_A}, {N_B} by {N_B}, and {N_A} by {N_B} kernels. "
        "This may take some time!"
```

(not checked line length, single line may be ok but can split across multiple lines like above if not)

**Review by jack89roberts — APPROVED** [review]
Added a minor formatting comment but LGTM. Re James' last comment - maybe a sanity check/test that `mmd` with `drop=0.8` is greater than with `drop=0.2` (for example), would be helpful just to convince that it is working as expected.
